easybuild/easyblocks/c/cudacompat.py

```diff
@@ -0,0 +1,143 @@
+        # or via 'accept_eula = True' in easyconfig file
+        self.check_accepted_eula(
+            name='NVIDIA-driver',
+            more_info='https://www.nvidia.com/content/DriverDownload-March2009/licence.php?lang=us'
+        )
+        super(EB_CUDAcompat, self).fetch_step(*args, **kwargs)
+
+    def extract_step(self):
+        """Extract the files without running the installer."""
+        execpath = self.src[0]['path']
+        tmpdir = os.path.join(self.builddir, 'tmp')
+        targetdir = os.path.join(self.builddir, 'extracted')
+        run_cmd("/bin/sh " + execpath + " --extract-only --tmpdir='%s' --target '%s'" % (tmpdir, targetdir))
+        self.src[0]['finalpath'] = targetdir
+
+    def install_step(self):
+        """Install CUDA compat libraries by copying library files and creating the symlinks."""
+        libdir = os.path.join(self.installdir, 'lib')
+        mkdir(libdir)
+
+        # From https://docs.nvidia.com/deploy/cuda-compatibility/index.html#installing-from-network-repo:
+        # The cuda-compat package consists of the following files:
+        #   - libcuda.so.* - the CUDA Driver
+        #   - libnvidia-nvvm.so.* - JIT LTO ( CUDA 11.5 and later only)
+        #   - libnvidia-ptxjitcompiler.so.* - the JIT (just-in-time) compiler for PTX files
+
+        library_globs = [
+            'libcuda.so.*',
+            'libnvidia-ptxjitcompiler.so.*',
+        ]
+        if LooseVersion(self.version) >= '11.5':
+            library_globs.append('libnvidia-nvvm.so.*')
+
+        startdir = self.cfg['start_dir']
+        nv_version = self.cfg['nv_version']
+        for library_glob in library_globs:
+            library_path = find_glob_pattern(os.path.join(startdir, library_glob))
+            library = os.path.basename(library_path)
+            # Sanity check the version
+            if library_glob == 'libcuda.so.*':
+                library_version = library.split('.', 2)[2]
+                if library_version != nv_version:
+                    raise EasyBuildError('Expected driver version %s (from nv_version) but found %s '
+                                         '(determined from file %s)', nv_version, library_version, library_path)
+
+            copy_file(library_path, os.path.join(libdir, library))
+            if library.endswith('.' + nv_version):
+                # E.g. libcuda.so.510.73.08 -> libcuda.so.1
+                versioned_symlink = library[:-len(nv_version)] + '1'
+            else:
+                # E.g. libnvidia-nvvm.so.4.0.0 -> libnvidia-nvvm.so.4
+                versioned_symlink = library.rsplit('.', 2)[0]
+            symlink(library, os.path.join(libdir, versioned_symlink), use_abspath_source=False)
+            # E.g. libcuda.so.1 -> libcuda.so
+            unversioned_symlink = versioned_symlink.rsplit('.', 1)[0]
+            symlink(versioned_symlink, os.path.join(libdir, unversioned_symlink), use_abspath_source=False)
+
+    def make_module_extra(self):
```

Comment:
We should store the driver version available at install time in the module file (and if there is no `nvidia-smi` we should probably require that they provide a version to check against in the easyblock).

The module file cannot be included in the EB rpath approach so I believe that we should clean out all the guessed paths (in particular the `lib` path should not be included in `LIBRARY_PATH`, see https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/scripts/rpath_args.py#L145). We also explicitly require the use of `LD_LIBRARY_PATH` for this, so this cannot be allowed to be omitted (which it frequently is when using rpath) so this should be set explicitly.

I'd suggest that we protect the use of `LD_LIBRARY_PATH` based on the driver version at load time, in particular if the driver version at load time is newer than `nv_version` then the path should _not_ be added to `LD_LIBRARY_PATH` and a warning should be issued that the package is (now) out of date. @bartoldeman mentioned that CC do something like this in the CUDA module itself:
```
if cuda_driver_library_available("11.7") == "compat" then
        depends_on("cuda-compat/.11.7")
end
```
You can get the driver version at load time with `nvidia-smi --query-gpu=driver_version --format=csv,noheader`. Of course it will be tricky to do this for tcl and lua.

Reply:
> We should store the driver version available at install time in the module file

Why? What would be gained by that, especially as we can't reliably get that anyway.

> The module file cannot be included in the EB rpath approach so I believe that we should clean out all the guessed paths

Makes sense to me.

> I'd suggest that we protect the use of LD_LIBRARY_PATH based on the driver version at load time

IMO this is the wrong place to do that. We already agreed that this will be a module to be manually installed and integrated into a deployment. So how this module is loaded is up to each site. We should only provide it for easy installation and not hold this up trying to anticipate different use cases. I mean it is possible that sites may add this to the CUDA module as a dependency either unconditionally or guarded by a condition such as that at CC. Or it may require (power) users to manually load the module, e.g. when it should be made explicit, that this is not really CUDA 11.6 but a compatibility CUDA 11.6

That it is module system dependent and hence hard to know generically if this module is to be loaded is another point against trying to do so and see it fail in some scenarios.

Reply:
> Why? What would be gained by that, especially as we can't reliably get that anyway.

We can reliably get it on nodes where `nvidia-smi`, my suggestion was that you get people to explicitly set it when `nvidia-smi` does not work. Even for nothing other than knowing what the driver was at install time, it is valuable information. But if the only use case is informational then ok, I can live without it.

> > I'd suggest that we protect the use of LD_LIBRARY_PATH based on the driver version at load time
> 
> IMO this is the wrong place to do that. We already agreed that this will be a module to be manually installed and integrated into a deployment. So how this module is loaded is up to each site. We should only provide it for easy installation and not hold this up trying to anticipate different use cases. I mean it is possible that sites may add this to the CUDA module as a dependency either unconditionally or guarded by a condition such as that at CC. Or it may require (power) users to manually load the module, e.g. when it should be made explicit, that this is not really CUDA 11.6 but a compatibility CUDA 11.6
> 
> That it is module system dependent and hence hard to know generically if this module is to be loaded is another point against trying to do so and see it fail in some scenarios.

Regardless of how the module is loaded, we know that loading it when the system driver is newer than `nv_version` will _not_ work. Since we know this, then we should encode it to make it easier for sites to diagnose problems.

Reply:
Running `nvidia-smi` for a module load is a bridge too far. It will be hard for sites to figure out what went wrong after a driver update though...